packages/flutter/lib/src/animation/curves.dart

```diff
@@ -267,6 +273,526 @@ class Cubic extends Curve {
   }
 }
 
+/// Abstract class that defines an API for evaluating 2D parametric curves.
+abstract class Curve2D {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const Curve2D();
+
+  /// Returns the position of the curve at point `t`.
+  ///
+  /// This function must ensure that the value of `t` is between 0.0 and
+  /// 1.0, inclusive.
+  ///
+  /// It is recommended that subclasses override [transformInternal] instead of
+  /// this function, as the above case is already enforced in the default
+  /// implementation of [transform], which delegates the remaining logic to
+  /// [transformInternal].
+  Offset transform(double t) {
+    assert(t >= 0.0 && t <= 1.0, 'parameter t is ${t.toStringAsFixed(2)}, which is outside of the range [0.0, 1.0]');
+    return transformInternal(t);
+  }
+
+  /// Returns the position of the curve at point `t`, where `t` is between 0.0
+  /// and 1.0, inclusive.
+  @protected
+  Offset transformInternal(double t) {
+    throw UnimplementedError();
+  }
+
+  @override
+  String toString() => '$runtimeType';
+}
+
+/// A class that holds a sample of a 2D curve, containing the X,Y coordinate of
+/// the curve at the parametric value T.
+///
+/// See also:
+///
+///  * [CatmullRomSpline.generateSamples], which generates samples of this type.
+class Curve2DSample {
+  /// A const constructor for the sample so that subclasses can be const.
+  ///
+  /// All arguments must not be null.
+  const Curve2DSample(this.t, this.value) : assert(t != null), assert(value != null);
+
+  /// The parametric value of this sample point.
+  final double t;
+
+  /// The location of the curve at parametric value [t].
+  final Offset value;
+
+  @override
+  String toString() {
+    return '[(${value.dx.toStringAsFixed(2)}, ${value.dy.toStringAsFixed(2)}), ${t.toStringAsFixed(2)}]';
+  }
+}
+
+/// A spline that passes smoothly through the given control points using a
+/// centripetal Catmull-Rom spline.
+///
+/// When the curve is evaluated with [transform], the output values will move
+/// smoothly from one control point to the next, passing through the control
+/// points.
+///
+/// {@template flutter.animation.curves.catmull_rom_description}
+/// Unlike most cubic splines, Catmull-Rom splines have the advantage that their
+/// curves pass through the control points given to them. They are cubic
+/// polynomial representations, and, in fact, Catmull-Rom splines can be
+/// converted mathematically into cubic splines. This class implements a
+/// "centripetal" Catmull-Rom spline implementation. The term centripetal
+/// implies that it won't form loops or self-intersections within a single
+/// segment.
+/// {@endtemplate}
+///
+/// See also:
+///
+///  * A Wikipedia article on [centripetal Catmull-Rom splines](https://en.wikipedia.org/wiki/Centripetal_Catmull%E2%80%93Rom_spline).
```

Comment:
uber-nit: Maybe phrase the two first bullet points as follows to be more in the style of Flutter's "see also"?
```
* [Centripetal Catmull–Rom splines](...) on Wikipedia.
* [Parameterization and Applications of Catmull-Rom Curves](...), a paper on using Catmull-Rom splies.
```